tasks/queue-granules/src/iterable.ts

```diff
@@ -0,0 +1,89 @@
+/**
+ * Group granules by collection and split into batches then split again on provider
+ * The purpose of this iterable is to avoid creating all the group and chunk arrays all
+ * at once to save heap space for large event inputs. This is done by using a sequence of
+ * generators that yield the chunks of granules one at a time.
+ *
+ * The *[Symbol.iterator] method loops over the collection looking for groups that have not
+ * already been yielded. Once the first granule of an unseen group is located, a new iterator is
+ * created that will search from that point in the collection of granules looking for other
+ * granules that belong to the same group.
+ *
+ * Finally a generator version of the _.chunk() function breaks the grouped granules into chunks.
+ */
+class GroupAndChunkIterable<
+  TElement,
+  TGroupProps extends Record<string, string | undefined | null> =
+  Record<string, string | undefined | null>,
```

Comment:
I think adding a default value for this type parameter makes the type parameter optional to specify and also makes it so that TypeScript can infer the type automatically when it needs to